platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/consensus/AncestorIterator.java

```diff
@@ -0,0 +1,210 @@
+ *
+ * <p>It is not threadsafe, and will silently fail without throwing any exceptions if you attempt to
+ * create and use two at the same time, even if they are in the same thread, and even if they are
+ * only used to read and not write. So always create only one at a time, and ensure that you are
+ * done with it before creating another one.
+ *
+ * <p>This returns all ancestors of the root event that are valid. It iterates in an order that
+ * always returns a parent before its child. The root itself is considered to be one of the
+ * ancestors, and is last in the iteration.
+ *
+ * <p>Recursion happens on self parents before other parents. So if there are multiple paths from
+ * the root to an event, it will use the path that stays on line of self parents for as far down as
+ * possible before leaving that line.
+ */
+public class AncestorIterator implements Iterator<EventImpl> {
+    private static final int INITIAL_STACK_SIZE = 300;
+    private final EventVisitedMark mark;
+    /** stack of EventImpl on the path to curr */
+    private final Deque<EventImpl> stackRef = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** stack of state */
+    private final Deque<IteratorState> stackState = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** stack of selfAncestor */
+    private final Deque<Boolean> stackSelfAncestor = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** stack of timeReachedRoot */
+    private final Deque<Instant> stackTime = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** the current event reached in the search */
+    private EventImpl curr;
+    /**
+     * a lambda which filters which ancestors are of interest: only each event e for which
+     * valid(e)==true
+     */
+    private Predicate<EventImpl> valid;
+    /**
+     * the time when the event last returned by the iterator's next() first reached the creator of
+     * root
+     */
+    private Instant timeReachedRoot;
+    /** becomes false when done and hasNext should return false */
+    private boolean hasNext = false;
+    /** the state of the state machine searching from curr */
+    private IteratorState state;
+    /** is curr a self ancestor of the judge? */
+    private boolean selfAncestor;
+
+    /**
+     * Create an iterator that will iterate over all ancestors of a given event
+     * @param mark the instance to use to mark events as visited
+     */
+    public AncestorIterator(@NonNull final EventVisitedMark mark) {
+        this.mark = mark;
+    }
+
+    /**
+     * The set of ancestors of the given event (the root of the search) for which valid is true.
+     * This will not include a valid ancestor that is only reachable through invalid ancestors.
+     *
+     * @param root the root event whose ancestors should be searched
+     * @param predicate do a depth-first search, but backtrack from any event e where
+     *     valid(e)==false
+     */
+    public void search(@NonNull final EventImpl root, @NonNull final Predicate<EventImpl> predicate) {
```

Comment:
Restarting the naming discussion. Can't remember what the suggestions were on the previous thread. How about `resetSearch()`?

Reply:
renamed to initializeSearch